alibi_detect/ad/model_distillation.py

```diff
@@ -0,0 +1,220 @@
+            Additional metrics whose progress will be displayed if verbose equals True.
+        callbacks
+            Callbacks used during training.
+        preprocess_fn
+            Preprocessing function applied to each training batch.
+        """
+        # train arguments
+        args = [self.distilled_model, loss_fn, X]
+        kwargs = {
+            'optimizer': optimizer,
+            'epochs': epochs,
+            'batch_size': batch_size,
+            'verbose': verbose,
+            'log_metric': log_metric,
+            'callbacks': callbacks,
+            'preprocess_fn': preprocess_fn,
+            'loss_fn_kwargs': {
+                'model': self.model,
+                'loss_type': self.loss_type,
+                'temperature': self.temperature
+            }
+        }
+
+        # train
+        trainer(*args, **kwargs)
+
+    def infer_threshold(self,
+                        X: np.ndarray,
+                        threshold_perc: float = 99.,
+                        margin: float = 0.,
+                        batch_size: int = int(1e10)
+                        ) -> None:
+        """
+        Update threshold by a value inferred from the percentage of instances considered to be
+        adversarial in a sample of the dataset.
+
+        Parameters
+        ----------
+        X
+            Batch of instances.
+        threshold_perc
+            Percentage of X considered to be normal based on the adversarial score.
+        margin
+            Add margin to threshold. Useful if adversarial instances have significantly higher scores and there
+            is no adversarial instance in X.
+        batch_size
+            Batch size used when computing scores.
+        """
+        # compute adversarial scores
+        adv_score = self.score(X, batch_size=batch_size)
+
+        # update threshold
+        self.threshold = np.percentile(adv_score, threshold_perc) + margin
+
+    def score(self, X: np.ndarray, batch_size: int = int(1e10), return_predictions: bool = False) \
+            -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray, np.ndarray]]:
+        """
+        Compute adversarial scores.
+
+        Parameters
+        ----------
+        X
+            Batch of instances to analyze.
+        batch_size
+            Batch size used when computing scores.
+        return_predictions
+            Whether to return the predictions of the classifier on the original and reconstructed instances.
+
+        Returns
+        -------
+        Array with adversarial scores for each instance in the batch.
+        """
+
+        # model predictions
+        y = predict_batch(self.model, X, batch_size=batch_size, proba=True)
+        y_distilled = predict_batch(self.distilled_model, X, batch_size=batch_size, proba=True)
+
+        # scale predictions
+        if self.temperature != 1.:
+            y = y ** (1 / self.temperature)  # type: ignore
+            y = y / tf.reshape(tf.reduce_sum(y, axis=-1), (-1, 1))
```

Comment:
Related to #138 : change to 
```python
y = (y / tf.reshape(tf.reduce_sum(y, axis=-1), (-1, 1))).numpy()
```
Otherwise returns mix of numpy arrays and tf tensors.